compiler/rustc_trait_selection/src/traits/object_safety.rs

```diff
@@ -802,22 +799,62 @@ fn contains_illegal_self_type_reference<'tcx>(
             }
         }
 
-        fn visit_const(&mut self, _c: &ty::Const<'tcx>) -> bool {
-            // FIXME(#72219) Look into the unevaluated constants for object safety violations.
-            // Do not walk substitutions of unevaluated consts, as they contain `Self`, even
-            // though the const expression doesn't necessary use it. Currently type variables
-            // inside array length expressions are forbidden, so they can't break the above
-            // rules.
-            false
+        fn visit_const(&mut self, ct: &ty::Const<'tcx>) -> bool {
+            // First check if the type of this constant references `Self`.
+            if self.visit_ty(ct.ty) {
```

Comment:
Can you also add a test for this? That will need the const generics feature gate so you can have non-usize constants, but that's fine.

Reply:
we already check this, `[u8; bar::<Self>]` only fails because of this bar is `Const { ty: fn bar with substs `Self`, ct: ZST }`.

If we did not add this `visit_ty` here we would never fail for constants, which took me a while to find out

Reply:
wait... what? as far as I can tell you only have `bar::<Self>()`, which is a constant of `usize` type. Even if its body contains other stuff.

Reply:
yeah, mean't `bar::<Self>()`. The issue here is that `bar::<Self>()` is a ZST of type `fn() -> usize { bar }`, so we have to look at the type of constants to see if they reference `Self`. The value of a const cannot reference `Self` afaik

Reply:
I get that, but I don't understand how the constant we are visiting is ever anything but a constant with `ty: usize`, where the constant's value may be `Unevaluated(DefIdOfbar, [Self])`, but that value is essentially irrelevant for the `visit_ty` here. Or is this during the visiting of the MIR of the `usize` typed constant?

Reply:
So we are visiting `[u8; bar::<Self>()]` and then the abstract const for the array length which is
```
[Leaf(Const { ty: fn() -> usize {bar::<Self>}, val: Value(Scalar(<ZST>)) }), FunctionCall(n0, [])]
```
So `Self` is only mentioned in the type of the function object.

Reply:
> During the visiting of the AbstractConst of the usize typed constant

Oh, now I realize where my confusion is coming from. It's actually both!

So...

```rust
struct Foo<T, const X: T>(T);
```

isn't possible, thus no

```
trait Foo {
    fn bar(&self) -> Foo<Self, something()>;
}
```

is possible either.